components/wifi/esp8266-driver/ESP8266/ESP8266.cpp

```diff
@@ -612,9 +630,24 @@ bool ESP8266::dns_lookup(const char *name, char *ip)
     return done;
 }
 
-nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
+nsapi_size_or_error_t ESP8266::send(int id, const void *data, uint32_t amount)
 {
+    if (_sock_i[id].proto == NSAPI_TCP) {
+        if (_sock_sending_id >= 0 && _sock_sending_id < SOCKET_COUNT) {
+            if (!_sock_i[id].send_fail) {
+                tr_debug("send(): Previous packet (socket %d) was not yet ACK-ed with SEND OK.", _sock_sending_id);
+                return NSAPI_ERROR_WOULD_BLOCK;
```

Comment:
What if modem doesn't reply with anything? Might `SEND OK` or `SEND FAIL` get lost? I'm just thinking out loud here so this isn't a request to change anything.

Reply:
This was discussed in the main thread. Please [see this comment](https://github.com/ARMmbed/mbed-os/pull/12157#issuecomment-570234338) and the previous one.
Long story short: `ERROR` is another possibility aside from `SEND OK` and `SEND FAIL`, but as experiments showed it is a recoverable error, so we basically ignore it.
If neither `SEND OK` nor `SEND FAIL` are coming we just keep returning `WOULD_BLOCK` to any new `send()` attempts. It's up to the application to decide how long this can be tolerated. I assume some socket timeout will take care of this in a typical mbed application?